### DISCUSSION TITLE
Trouble with libpng on M1

### DISCUSSION THREAD

**Discussion by EllipticEllipsis** [Homebrew/discussions#2625] [Getting started] [answered]
I installed libpng with brew. I'm not very familiar with brew, so I've probably done something stupid accidentally. 
- It won't find the header when trying to compile with it without doing this: `-I /opt/homebrew/include/`, and
- Even if I do that, and and link it manually with `-L /opt/homebrew/lib/` (or anything similar: I don't know what the right directory actually is), it still says

```
Undefined symbols for architecture arm64:
  "_png_create_info_struct", referenced from:
      ImageBackend::ReadPng(char const*) in ImageBackend.o
      ImageBackend::WritePng(char const*) in ImageBackend.o
  "_png_create_read_struct", referenced from:
      ImageBackend::ReadPng(char const*) in ImageBackend.o
  "_png_create_write_struct", referenced from:
      ImageBackend::WritePng(char const*) in ImageBackend.o
  "_png_destroy_read_struct", referenced from:
      ImageBackend::ReadPng(char const*) in ImageBackend.o
  "_png_destroy_write_struct", referenced from:
      ImageBackend::WritePng(char const*) in ImageBackend.o
  "_png_get_bit_depth", referenced from:
      ImageBackend::ReadPng(char const*) in ImageBackend.o
  "_png_get_color_type", referenced from:
      ImageBackend::ReadPng(char const*) in ImageBackend.o
  "_png_get_image_height", referenced from:
      ImageBackend::ReadPng(char const*) in ImageBackend.o
  "_png_get_image_width", referenced from:
      ImageBackend::ReadPng(char const*) in ImageBackend.o
  "_png_get_rowbytes", referenced from:
      ImageBackend::ReadPng(char const*) in ImageBackend.o
  "_png_init_io", referenced from:
      ImageBackend::ReadPng(char const*) in ImageBackend.o
      ImageBackend::WritePng(char const*) in ImageBackend.o
  "_png_read_image", referenced from:
      ImageBackend::ReadPng(char const*) in ImageBackend.o
  "_png_read_info", referenced from:
      ImageBackend::ReadPng(char const*) in ImageBackend.o
  "_png_read_update_info", referenced from:
      ImageBackend::ReadPng(char const*) in ImageBackend.o
  "_png_set_IHDR", referenced from:
      ImageBackend::WritePng(char const*) in ImageBackend.o
  "_png_set_PLTE", referenced from:
      ImageBackend::WritePng(char const*) in ImageBackend.o
  "_png_set_expand_gray_1_2_4_to_8", referenced from:
      ImageBackend::ReadPng(char const*) in ImageBackend.o
  "_png_set_gray_to_rgb", referenced from:
      ImageBackend::ReadPng(char const*) in ImageBackend.o
  "_png_set_longjmp_fn", referenced from:
      ImageBackend::ReadPng(char const*) in ImageBackend.o
      ImageBackend::WritePng(char const*) in ImageBackend.o
  "_png_set_strip_16", referenced from:
      ImageBackend::ReadPng(char const*) in ImageBackend.o
  "_png_set_tRNS", referenced from:
      ImageBackend::WritePng(char const*) in ImageBackend.o
  "_png_write_end", referenced from:
      ImageBackend::WritePng(char const*) in ImageBackend.o
  "_png_write_image", referenced from:
      ImageBackend::WritePng(char const*) in ImageBackend.o
  "_png_write_info", referenced from:
      ImageBackend::WritePng(char const*) in ImageBackend.o
ld: symbol(s) not found for architecture arm64
```

I checked, and the archive really is compiled for arm64:
```
$ lipo -info /opt/homebrew/lib/libpng.a
Non-fat file: /opt/homebrew/lib/libpng.a is architecture: arm64
```

so I'm at a loss.

**Comment by gromgit** [comment, marked as answer]
This looks more like "novice pilot error" than something actually broken. Specifically:

> * It won't find the header when trying to compile with it without doing this: `-I /opt/homebrew/include/`

This is standard C practice for pointing the compiler to non-system include directories.

> * Even if I do that, and and link it manually with `-L /opt/homebrew/lib/`

That just tells the compiler where to look for the PNG library, but you didn't tell the compiler to actually link it. For that, you need to add `-lpng` to link with the dynamic library (`libpng.dylib`). If you actually wanted to import the static library in your executable, add `/opt/homebrew/lib/libpng.a` instead.

**Reply by EllipticEllipsis** [reply]
Thank you for your reply.

> This is standard C practice for pointing the compiler to non-system include directories.

True, but it did not have this problem when compiled to x86 (which I did by accident from having the terminal in Rosetta to start with), so naturally I assumed the same would be true of ARM.

> 
> > * Even if I do that, and and link it manually with `-L /opt/homebrew/lib/`
> 
> That just tells the compiler where to look for the PNG library, but you didn't tell the compiler to actually link it. For that, you need to add `-lpng` to link with the dynamic library (`libpng.dylib`). If you actually wanted to import the static library in your executable, add `/opt/homebrew/lib/libpng.a` instead.

`-lpng` by itself gives `ld: library not found for -lpng`, so I'm not sure what you're referring to here.

**Reply by EllipticEllipsis** [reply]
And linking the static library manually gives exactly the same error as my original post.

**Reply by EllipticEllipsis** [reply]
Actually, I tell a lie... it's a bit different:
```
Undefined symbols for architecture arm64:
  "_adler32", referenced from:
      _png_icc_set_sRGB in libpng.a(png.o)
  "_crc32", referenced from:
      _png_reset_crc in libpng.a(png.o)
      _png_calculate_crc in libpng.a(png.o)
      _png_icc_set_sRGB in libpng.a(png.o)
  "_deflate", referenced from:
      _png_compress_IDAT in libpng.a(pngwutil.o)
      _png_text_compress in libpng.a(pngwutil.o)
  "_deflateEnd", referenced from:
      _png_destroy_write_struct in libpng.a(pngwrite.o)
      _png_deflate_claim in libpng.a(pngwutil.o)
  "_deflateInit2_", referenced from:
      _png_deflate_claim in libpng.a(pngwutil.o)
  "_deflateReset", referenced from:
      _png_deflate_claim in libpng.a(pngwutil.o)
  "_inflate", referenced from:
      _png_zlib_inflate in libpng.a(pngrutil.o)
     (maybe you meant: _png_zlib_inflate)
  "_inflateEnd", referenced from:
      _png_destroy_read_struct in libpng.a(pngread.o)
  "_inflateInit2_", referenced from:
      _png_inflate_claim in libpng.a(pngrutil.o)
  "_inflateReset", referenced from:
      _png_reset_zstream in libpng.a(png.o)
      _png_decompress_chunk in libpng.a(pngrutil.o)
  "_inflateReset2", referenced from:
      _png_inflate_claim in libpng.a(pngrutil.o)
  "_inflateValidate", referenced from:
      _png_inflate_claim in libpng.a(pngrutil.o)
ld: symbol(s) not found for architecture arm64
```

**Reply by gromgit** [reply]
> > This is standard C practice for pointing the compiler to non-system include directories.
> 
> True, but it did not have this problem when compiled to x86 (which I did by accident from having the terminal in Rosetta to start with), so naturally I assumed the same would be true of ARM.

x86 Homebrew installs to `/usr/local`, which happens to be the _de facto_ "local software" hierarchy across most of Unix-dom, so many projects automatically add `-I/usr/local/include` to their configuration.

> > That just tells the compiler where to look for the PNG library, but you didn't tell the compiler to actually link it. For that, you need to add `-lpng` to link with the dynamic library (`libpng.dylib`). If you actually wanted to import the static library in your executable, add `/opt/homebrew/lib/libpng.a` instead.
> 
> `-lpng` by itself gives `ld: library not found for -lpng`, so I'm not sure what you're referring to here.

As I said, you need to _add_ one to the other, i.e. `-L/opt/homebrew/lib -lpng`. The first adds an additional directory to your compiler's library search path, the second tells it which (dynamic) library to link.

**Reply by EllipticEllipsis** [reply]
Oh, I think I had the wrong folder anyway. Thank you very much, you're a lifesaver!
